ykgenerate.1
.\" Copyright (c) 2013 Yubico AB
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\"   * Redistributions of source code must retain the above copyright
.\"      notice, this list of conditions and the following disclaimer.
.\"
.\"   * Redistributions in binary form must reproduce the above
.\"      copyright notice, this list of conditions and the following
.\"      disclaimer in the documentation and/or other materials provided
.\"      with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.TH YKGENERATE 1 "October 2013" yubico\-c "User Commands"
.SH NAME
ykgenerate \- construct and encrypt an OTP from given data
.SH SYNOPSIS
.B ykgenerate
\fIAESKEY YK_INTERNALNAME YK_COUNTER YK_LOW YK_HIGH YK_USE\fR
.SH DESCRIPTION
.PP
The ykgenerate tool generate OTPs by encrypting given values under the
given encryption key.
.SH OPTIONS
.TP
\fBAESKEY\fR
Hex encoded AES\-key.
.TP
\fBYK_INTERNALNAME\fR
Hex encoded internal identity (48 bit).
.TP
\fBYK_COUNTER\fR
Hex encoded counter (16 bit).
.TP
\fBYK_LOW\fR
Hex encoded timestamp low (16 bit).
.TP
\fBYK_HIGH\fR
Hex encoded timestamp high (8bit).
.TP
\fBYK_USE\fR
Hex encoded use (8 bit).
.SH EXAMPLES
Typical usage of the tool is as follows (be careful with linewrapping
of the manpage).
.RS
ykgenerate ecde18dbe76fbd0c33330f1c354871db 8792ebfe26cc 0013 c0a8 00 10
.RE
The expected output is the string "hcfktknicvbcnhbkvigcfhgddhhhrknc".
.SH "SEE ALSO"
YubiKeys can be obtained from Yubico: http://www.yubico.com/


NEWS
Libyubikey NEWS -- History of user-visible changes.             -*- outline -*-

* Version 1.12 (unreleased)

* Version 1.11 (released 2013-10-18)

** Add man pages.

** Self-tests may be run under valgrind to catch bugs.
Disable with "./configure --disable-valgrind-tests".

* Version 1.10 (released 2013-03-12)

** Added ./configure --enable-gcc-warnings to enable a lot of warnings.

** Warning fixes, build fixes etc.

* Version 1.9 (released 2012-05-31)

** Updated ld-version-script from gnulib to silence warnings.

** Fix out-of-tree builds.

* Version 1.8 (released 2012-03-13)

** modhex tool: Fix memory leak.

** Fix insecure rpath.

* Version 1.7 (released 2010-11-24)

** yubikey.h: Possible to use from C++ using extern namespace scoping.
Reported by Giampaolo Bellini <iw2lsi@gmail.com>.

* Version 1.6 (released 2010-08-03)

** New API to generate OTPs.
The API is yubikey_generate.  An helper function yubikey_aes_encrypt
was added too.  Added YUBIKEY_OTP_SIZE constant for OTP buffer size.

** ykgenerate: New tool to generate OTPs.

** ykdebug/ykparse: The old tool "ykdebug" has been renamed to "ykparse".

* Version 1.5 (released 2008-09-31)

** The library now use a linker version script.
It falls back to using Libtool's -export-symbols-regex if linker
version scripts are not supported.

** Added new APIs to check validity of hex/modhex strings.
The APIs are yubikey_modhex_p and yubikey_hex_p.

* Version 1.4 (released 2009-04-30)

** Improved instructions in README on how to build the package.

** Add interfaces yubikey_hex_encode and yubikey_hex_decode.

** The ykdebug tool now takes hex encoded AES keys instead of modhex encoded.

* Version 1.3 (released 2009-03-25)

** Add YUBIKEY_MODHEX_MAP.

** Use char* instead of uint8_t* for modhex functions.

* Version 1.2 (released 2009-01-13)

** Don't link to libpam.

** Add libtool -export-symbols-regex to restrict exported symbols.

* Version 1.1 (released 2008-06-10)

** Include simple.mk in distribution.

* Version 1.0 (released 2008-06-10)

** Uses autoconf, automake and libtool.

** Renamed function namespace from pof_* to yubikey_*.


ykgenerate.c
/* ykgenerate.c --- Example command line interface for generating OTPs.
 *
 * Written by Olov Danielson <olov@yubico.com>.
 * Copyright (c) 2006-2012 Yubico AB
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *    * Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *
 *    * Redistributions in binary form must reproduce the above
 *      copyright notice, this list of conditions and the following
 *      disclaimer in the documentation and/or other materials provided
 *      with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include "yubikey.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>
#include <limits.h>

int
main (int argc, char *argv[])
{
  uint8_t key[YUBIKEY_KEY_SIZE];
  char otp[YUBIKEY_OTP_SIZE];
  char *aeskey, *yk_internalname, *yk_counter, *yk_low, *yk_high, *yk_use;
  yubikey_token_st tok;

  /* Initiate pseudo-random generator */
  srand (time (NULL));

  /* Parse command line parameters. */
  if (argc < 7)
    {
      printf ("Usage: %s <aeskey> <yk_internalname> <yk_counter> <yk_low> "
	      "<yk_high> <yk_use>\n", argv[0]);
      printf (" AESKEY:\t\tHex encoded AES-key.\n");
      printf (" YK_INTERNALNAME:\tHex encoded yk_internalname (48 bit).\n");
      printf (" YK_COUNTER:\t\tHex encoded counter (16 bit).\n");
      printf (" YK_LOW:\t\tHex encoded timestamp low (16 bit).\n");
      printf (" YK_HIGH:\t\tHex encoded timestamp high (8bit).\n");
      printf (" YK_USE:\t\tHex encoded use (8 bit).\n");
      return EXIT_FAILURE;
    }

  aeskey = argv[1];
  if (strlen (aeskey) != 32)
    {
      printf ("error: Hex encoded AES-key must be 32 characters.\n");
      return EXIT_FAILURE;
    }

  yk_internalname = argv[2];
  if (strlen (yk_internalname) != 12)
    {
      printf ("error: Hex encoded yk_internalname must be 12 characters.\n");
      return EXIT_FAILURE;
    }

  yk_counter = argv[3];
  if (strlen (yk_counter) != 4)
    {
      printf ("error: Hex encoded yk_counter must be 4 characters.\n");
      return EXIT_FAILURE;
    }

  yk_low = argv[4];
  if (strlen (yk_low) != 4)
    {
      printf ("error: Hex encoded yk_low must be 4 characters.\n");
      return EXIT_FAILURE;
    }

  yk_high = argv[5];
  if (strlen (yk_high) != 2)
    {
      printf ("error: Hex encoded yk_high must be 2 characters.\n");
      return EXIT_FAILURE;
    }

  yk_use = argv[6];
  if (strlen (yk_use) != 2)
    {
      printf ("error: Hex encoded yk_use must be 2 characters.\n");
      return EXIT_FAILURE;
    }

  yubikey_hex_decode ((char *) key, aeskey, YUBIKEY_KEY_SIZE);

  /* Fill upp tok with values */
  yubikey_hex_decode ((char *) &tok.uid, yk_internalname, 6);
  {
    unsigned long i;
    char *endptr;

    i = strtoul (yk_counter, &endptr, 16);
    if (*endptr != '\0' || i == ULONG_MAX)
      {
	printf ("error: Hex encoded yk_counter must be 4 hex characters.\n");
	return EXIT_FAILURE;
      }
    tok.ctr = i;

    i = strtoul (yk_low, &endptr, 16);
    if (*endptr != '\0' || i == ULONG_MAX)
      {
	printf ("error: Hex encoded yk_low must be 4 hex characters.\n");
	return EXIT_FAILURE;
      }
    tok.tstpl = i;
  }
  yubikey_hex_decode ((char *) &tok.tstph, yk_high, 1);
  yubikey_hex_decode ((char *) &tok.use, yk_use, 1);
  tok.rnd = rand ();
  tok.crc = ~yubikey_crc16 ((void *) &tok, sizeof (tok) - sizeof (tok.crc));

  yubikey_generate ((void *) &tok, key, otp);

  {
    size_t i;

    for (i = 0; i < 32; i++)
      printf ("%c", otp[i]);

    printf ("\n");
  }

  return EXIT_SUCCESS;
}
